nm: rewrite profile modification code

cathay4t opened the pull request:
Top design:

* NmProfiles - Handling actions required the knowledge of multiple iface.
* NmProfile - Present a NM profile of single interface.
* ProfileAdd - Profile addition.
* ProfileUpdate - Profile update.
* ProfileDelete - Profile deletion.
* DeviceReapply - Device configuration reapply.
* DeviceDelete - Device deletion.
* ProfileActivation - Profile Activation.
  Even this is related to NM.RemoteConnection, but most of it code
  is relating to NM.ActiveConnection, hence placed in
  `nm/activate_connection.py`
  Has `run()` for normal activation based on profile.
  Has `wait()` for waiting activation used by checkpoint rollback.
* ActiveConnectionDeactivate - Deactivation
* `create_new_nm_simple_conn()` - Create SimpleConnection



Group activation:
* The NmProfile has `_gen_actions()` which determine which action
  groups should this profile been tagged.
* The NmProfiles will iterate `NmProfile.ACTIONS` in the predefined order,
  any profile having that action will be invoked by
  do_action(action).
* With this design, group action is solely determined by NmProfile
  itself.

Removed duplicate test cases in `tests/integration/nm` and only keep
that folder holding NM specific integration test cases.

cathay4t commented:
let's wait PR #1408

ffmancera commented on libnmstate/nm/active_connection.py:
```diff
@@ -20,162 +20,369 @@
+
+    @staticmethod
+    def wait(ctx, nm_ac, nm_dev):
+        obj = ProfileActivation(
+            ctx,
+            nm_dev.get_iface(),
+            get_iface_type(nm_dev),
+            None,
+            nm_dev,
+        )
+        obj._nm_ac = nm_ac
+        obj._action = (
+            f"Waiting activation of {obj._iface_name} {obj._iface_type}"
+        )
+        ctx.register_async(obj._action)
+        obj._wait_profile_activation()
```
What about using `activation_handler` or a better name instead of `obj`? Thanks!

cathay4t replied:
Changed to `activation`, even through I still perfer `obj`. :)

ffmancera replied:
Thanks!

ffmancera commented on libnmstate/nm/active_connection.py:
```diff
@@ -20,162 +20,369 @@
+            if self._nm_dev:
+                self._wait_nm_dev_activation()
+            if not self._nm_ac and not self._nm_dev:
+                self._ctx.fail(
+                    NmstateInternalError(
+                        f"{self._action} failed: no nm_ac or nm_dev"
+                    )
+                )
```
I think this is missing `self._activation_clean_up` before failing, maybe there could be a condition race when rolling back. Not sure about it.

cathay4t replied:
There is no `nm_ac` or `nm_dev`, nothing need to be clean up.

ffmancera commented on libnmstate/nm/context.py:
```diff
@@ -208,13 +212,12 @@ def wait_all_finish(self):
         if self._error:
             # The queue and error should be flush and perpare for another run
+            self._cancellable.cancel()
+            self.refresh_content()
```
After this patch I will investigate if `refresh_content()` is still needed.

cathay4t replied:
That is trackked at https://github.com/nmstate/nmstate/issues/1398

ffmancera reviewed (changes requested):
In general it looks good to me. I've made some comments below. In addition, why is this patch removing a lot of nm integration tests? Thanks!

cathay4t commented:
The reason I remove a lot NM integreation test case:

```
Removed duplicate test cases in tests/integration/nm and only keep
that folder holding NM specific integration test cases.
```

Many old nm test case is just duplicated effort on top level interrogation test cases, there is no need to keep them maintained.

cathay4t committed:
nm: rewrite profile modification code

Top design:
    * NmProfiles - Handling actions required the knowledge of multiple iface.
    * NmProfile - Present a NM profile of single interface.
    * ProfileAdd - Profile addition.
    * ProfileUpdate - Profile update.
    * ProfileDelete - Profile deletion.
    * DeviceReapply - Device configuration reapply.
    * DeviceDelete - Device deletion.
    * ProfileActivation - Profile Activation.
      Even this is related to NM.RemoteConnection, but most of it code
      is relating to NM.ActiveConnection, hence placed in
      `nm/activate_connection.py`
      Has `run()` for normal activation based on profile.
      Has `wait()` for waiting activation used by checkpoint rollback.
    * ActiveConnectionDeactivate - Deactivation
    * `create_new_nm_simple_conn()` - Create SimpleConnection

Group activation:
    * The NmProfile has `_gen_actions()` which determine which action
      groups should this profile been tagged.
    * The NmProfiles will iterate `NmProfile.ACTIONS` in the predefined order,
      any profile having that action will be invoked by
      do_action(action).
    * With this design, group action is solely determined by NmProfile
      itself.

Removed duplicate test cases in `tests/integration/nm` and only keep
that folder holding NM specific integration test cases.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
wait PR #1412

ffmancera reviewed (approved):
LGTM. Thanks!
